lib/ddtrace/contrib/rack/middlewares.rb

```diff
@@ -113,18 +101,14 @@ def call(env)
             request_span.status = 1
           end
 
-          request_span.finish()
-
-          [status, headers, response]
-        end
-
-        private
+          # ensure the request_span is finished and the context reset;
+          # this assumes that the Rack middleware creates a root span
+          request_span.finish
 
-        # TODO: Remove this once we change how context propagation works. This
-        # ensures we clean thread-local variables on each HTTP request avoiding
-        # memory leaks.
-        def clean_context
-          @tracer.provider.context = Datadog::Context.new
+          # TODO: Remove this once we change how context propagation works. This
+          # ensures we clean thread-local variables on each HTTP request avoiding
+          # memory leaks.
+          tracer.provider.context = Datadog::Context.new
```

Comment:
removed a level of indirection

Reply:
also we should clean the `Context` at the end of the request, since it's possible another layer (i.e. Puma/Unicorn instrumentation?) creates a new `Context` before this middleware.